util-flywayinstaller/src/main/java/com/github/cafapi/util/flywayinstaller/Migrator.java

```diff
@@ -0,0 +1,117 @@
+    private static final String CONNECTION_URL_REGEX = "^.*\\/\\/.+:\\d+\\/";
+
+    private Migrator()
+    {
+    }
+
+    public static void migrate(
+        final boolean allowDBDeletion,
+        final String connectionString,
+        final String dbName,
+        final String username,
+        final String password
+    ) throws InvalidConnectionStringException, SQLException
+    {
+        LOGGER.debug("Arguments received"
+            + " allowDBDeletion: {}\n"
+            + " connectionString: {}\n"
+            + " dbName: {}\n"
+            + " username: {}\n"
+            + " password: {}", allowDBDeletion, connectionString, dbName, username, password);
+
+        LOGGER.info("Starting migration ...");
+        try (final BasicDataSource dbSource = new BasicDataSource()) {
+            dbSource.setUrl(checkAndConvertConnectionUrl(connectionString));
+            dbSource.setUsername(username);
+            dbSource.setPassword(password);
+
+            final boolean exists = checkDBExists(dbSource, dbName);
+            LOGGER.debug("Database exists: {}", exists);
+            if (!exists || allowDBDeletion) {
+                resetOrCreateDatabase(dbSource, exists, dbName);
+            }
+
+            LOGGER.info("About to perform DB update.");
+            final Flyway flyway = Flyway.configure()
+                .dataSource(dbSource.getUrl() + dbName, username, password)
+                .baselineOnMigrate(true)
+                .load();
+            flyway.migrate();
+            flyway.validate();
+            LOGGER.info("DB update finished.");
+        }
+        LOGGER.info("Migration completed ...");
+    }
+
+    private static void resetOrCreateDatabase(
+        final BasicDataSource dbSource,
+        final boolean exists,
+        final String dbName) throws SQLException
+    {
+        final String savedUrl = dbSource.getUrl();
+        dbSource.setUrl(dbSource.getUrl() + "postgres");
+        LOGGER.debug(dbSource.getUrl());
+        try (final Connection connection = dbSource.getConnection();
+             final Statement statement = connection.createStatement()) {
+            if (exists) {
+                LOGGER.info("force deletion has been specified.\nDeleting database {}", dbName);
+                statement.executeUpdate("DROP DATABASE " + dbName);
+                LOGGER.info("DELETED database: {}", dbName);
+            }
+            statement.executeUpdate("CREATE DATABASE " + dbName);
+            LOGGER.info("Created new database: {}", dbName);
+        }
+        dbSource.setUrl(savedUrl);
+    }
+
+    private static boolean checkDBExists(final BasicDataSource dbSource, final String dbName) throws SQLException
+    {
+        try (final Connection connection = dbSource.getConnection();
+             final Statement statement = connection.createStatement()) {
+            final boolean exists = statement.executeQuery("SELECT * FROM pg_database WHERE datname=lower('" + dbName + "');").next();
```

Comment:
- What if `dbName` contained a `'`?  This is classic SQL injection.  You should be using a parameterized query.
- Is this definitely a safe way to check if the database exists?  I presume the `pg_database` is cluster-wide and so I'm wondering if super-user permissions are required to check this?  Although I suppose they would be required anyway for `CREATE DATABASE` so it's probably ok.

Reply:
https://stackoverflow.com/a/26582770/10218144 It seems we can't use the preparedStatement for drop and create as we don't want to have 'dbName'.
If we received a dbName containing a ' like
`CREATE DATABASE 'test';`
then it would fail with:
`SQL Error [42601]: ERROR: syntax error at or near "'test'"`